### PR TITLE
Fit deaths by age

### PR DESCRIPTION
**Pull request by pabloperguz**
This PR implements fitting to hospital deaths by age. It requires spimalot 0.12.28 (please not tests will fail until the release is available).

In this PR:
- New deaths by age data is processed for the particle filter
- Remains of the old processing for comparing deaths by age are scrapped
- Newly outputted (fitted) trajectories are plotted against data

### CODE REVIEW

File: R/pars.R [hunk 1]
```diff
@@ -188,6 +188,8 @@ spim_pars_beta <- function(date, last_beta_days_ago = 21) {
   ##                  (help the model stabilise a long period of time)
   ## 28. 2021-10-01 - Point before recent increase in cases - hospitalisations
   ## 29. 2021-10-22 - School half-term - Point before recent plateau in cases
+  ## 30. 2021-11-01 - Schools return
+  ## 31. 2021-12-06 - Pragmatic date - Omicron transmission increases
   ## 30. Date - last_beta_days_ago
```
Comment (richfitz) [inline R/pars.R:193]: these numbers now out of sequence - please increase 30. on line 193 to 32

File: R/fit_process.R [hunk 2]
```diff
@@ -312,78 +303,6 @@ calculate_lancelot_ifr_t <- function(samples, multistrain) {
 
 ## All the functions below here have awful names, but none are
 ## exported so we can tidy this up later.
-extract_outputs_by_age <- function(sample, what) {
```
Comment (richfitz) [inline R/fit_process.R:315]: Always nice to see old code get deleted :)

**Review by richfitz — COMMENTED** [review]
I believe this PR will fail until the sircovid one has merged and this one is rerun.

File: R/data.R [hunk 3]
```diff
@@ -78,7 +78,10 @@ spim_data_single <- function(date, region, model_type, rtm, serology,
   ## are too likely to be back-filled to be reliable
   i <- seq(to = nrow(data), length.out = trim_deaths)
   data[i, c("deaths", "deaths_hosp", "deaths_comm",
-            "deaths_carehomes",  "deaths_non_hosp")] <- NA
+            "deaths_carehomes",  "deaths_non_hosp",
```
Comment (edknock) [inline R/data.R:81]: might just be easier to use a grep here?

File: R/data.R [hunk 4]
```diff
@@ -168,13 +177,18 @@ spim_lancelot_data_rtm <- function(date, region, model_type, data, full_data) {
...
     data$deaths <- data$death2
+    data[, deaths_hosp_age] <- NA_integer_
     data$deaths_hosp <- NA_integer_
     data$deaths_comm <- NA_integer_
     data$deaths_carehomes <- NA_integer_
     data$deaths_non_hosp <- NA_integer_
   } else {
     data$deaths <- NA_integer_
-    data$deaths_hosp <- data$death3
+    if (all(!is.na(data[, deaths_hosp_age])) && !full_data) {
```
Comment (edknock) [inline R/data.R:187]: I think this would be better off moved to the `if(!full_data)` section after `ret <- data.frame(...`

File: R/data.R (hunk 3 above)
Comment (richfitz) [reply R/data.R:81]: Not a huge fan of grep for these things, tbh

File: R/fit_process.R [hunk 5]
```diff
@@ -53,14 +53,7 @@ spim_fit_process <- function(samples, parameters, data, control,
...
 
-  ## TODO: currently still used to compare outputs of deaths by age (not fitted)
-  ## this will be removed soon, when we introduce fitting to deaths by age
-  message("Summarising deaths")
-  deaths <- extract_outputs_by_age(samples_thin, "D_hosp") # slow
-  i_deaths_data <- colnames(deaths$output_t)
-  deaths$data <- data$rtm[data$rtm$region == region,
-                          c("date", "region", i_deaths_data)]
-  deaths$data[is.na(deaths$data)] <- 0
+  ## TODO: process D_hosp states by age before trajectories are reduced below
```
Comment (edknock) [inline R/fit_process.R:56]: Is this comment needed anymore?

File: R/pars.R [hunk 6]
```diff
@@ -188,6 +188,8 @@ spim_pars_beta <- function(date, last_beta_days_ago = 21) {
   ##                  (help the model stabilise a long period of time)
   ## 28. 2021-10-01 - Point before recent increase in cases - hospitalisations
   ## 29. 2021-10-22 - School half-term - Point before recent plateau in cases
+  ## 30. 2021-11-01 - Schools return
+  ## 31. 2021-12-06 - Pragmatic date - Omicron transmission increases
```
Comment (edknock) [inline R/pars.R:192]: What's the logic of this date?

Comment (edknock) [reply R/pars.R:192]: Would 2021-12-08 be better (announcement of move to Plan B)? https://www.gov.uk/government/news/prime-minister-confirms-move-to-plan-b-in-england

File: R/fit_process.R (hunk 2 above)
Comment (pabloperguz) [reply R/fit_process.R:315]: Absolutely liberating!

File: R/data.R (hunk 3 above)
Comment (pabloperguz) [reply R/data.R:81]: I can see the benefit of both approaches. How about this? Makes it easier to modify, should we decide to change the age bands, but still very explicit as to what we are NAing

`deaths_hosp_age <- paste0("deaths_hosp_", c(0, seq(50, 80, 5)),
                            "_", c(seq(49, 79, 5), 120))
  deaths_hosp_age <- gsub("120", "plus", deaths_hosp_age)
  i <- seq(to = nrow(data), length.out = trim_deaths)
  data[i, c("deaths", "deaths_hosp", "deaths_comm",
            "deaths_carehomes",  "deaths_non_hosp",
            deaths_hosp_age)] <- NA`

Comment (edknock) [reply R/data.R:81]: Happy with doing it that way instead!

File: R/pars.R (hunk 6 above)
Comment (pabloperguz) [reply R/pars.R:192]: Good shout!

File: R/data.R [hunk 7]
```diff
@@ -449,6 +468,12 @@ spim_lancelot_data_rtm <- function(date, region, model_type, data, full_data) {
     ret$pillar2_under15_pos <- NA_integer_
     ret$pillar2_under15_cases <- NA_integer_
 
+    ## Do not fit to aggregated hospital deaths
+    deaths_hosp_age <- gsub("death", "deaths_hosp", ret)
+    if (all(!is.na(ret[, deaths_hosp_age]))) {
```
Comment (edknock) [inline R/data.R:473]: `any` instead of `all`?

File: R/data.R [hunk 8]
```diff
@@ -449,6 +468,12 @@ spim_lancelot_data_rtm <- function(date, region, model_type, data, full_data) {
     ret$pillar2_under15_pos <- NA_integer_
     ret$pillar2_under15_cases <- NA_integer_
 
+    ## Do not fit to aggregated hospital deaths
+    deaths_hosp_age <- gsub("death", "deaths_hosp", ret)
```
Comment (edknock) [inline R/data.R:472]:
```suggestion
    deaths_hosp_age <- gsub("death", "deaths_hosp", deaths_hosp_age)
```

**Review by edknock — APPROVED** [review]

**Pull request merged**